base/compiler/abstractinterpretation.jl

```diff
@@ -571,17 +605,75 @@ struct MethodCallResult
     edgecycle::Bool
     edgelimited::Bool
     edge::Union{Nothing,MethodInstance}
+    edge_effects::Effects
     function MethodCallResult(@nospecialize(rt),
                               edgecycle::Bool,
                               edgelimited::Bool,
-                              edge::Union{Nothing,MethodInstance})
-        return new(rt, edgecycle, edgelimited, edge)
+                              edge::Union{Nothing,MethodInstance},
+                              edge_effects::Effects)
+        return new(rt, edgecycle, edgelimited, edge, edge_effects)
+    end
+end
+
+function is_all_const_arg((; fargs, argtypes)::ArgInfo)
+    for a in argtypes
+        if !isa(a, Const) && !isconstType(a) && !issingletontype(a)
+            return false
+        end
     end
+    return true
+end
+
+function pure_eval_call_const_args(interp::AbstractInterpreter,
```

Comment:
"pure_eval_call_const_args" sounds a bit confusing, especially because we use `_call_in_world_nonpure` here.
(and also we are already having `@pure`, but this one is really separated from that "pure" terminology)